File: src/tools/registry.ts
/**
 * ToolRegistry - Manages available tools and capabilities
 *
 * This class provides a registry for all available tools that can be
 * called through the ACP protocol, including filesystem, terminal, and
 * cursor-specific tools.
 */

import type { ToolKind, ToolCallLocation } from '@agentclientprotocol/sdk';
import {
  ToolError,
  type AdapterConfig,
  type Logger,
  type Tool,
  type ToolProvider,
  type ToolCall,
  type ToolResult,
} from '../types';
import { FilesystemToolProvider } from './filesystem';
import { TerminalToolProvider } from './terminal';
import { CursorToolsProvider } from './cursor-tools';
import type { ToolCallManager } from './tool-call-manager';

export class ToolRegistry {
  private config: AdapterConfig;
  private logger: Logger;
  private providers = new Map<string, ToolProvider>();
  private tools = new Map<string, Tool>();
  private toolCallManager?: ToolCallManager;

  constructor(config: AdapterConfig, logger: Logger) {
    this.config = config;
    this.logger = logger;

    this.logger.debug('ToolRegistry initialized');

    // Initialize built-in tool providers
    this.initializeProviders();
  }

  /**
   * Set the tool call manager for reporting tool execution
   */
  setToolCallManager(manager: ToolCallManager): void {
    this.toolCallManager = manager;
    this.logger.debug('ToolCallManager registered with ToolRegistry');
  }

  /**
   * Registers a tool provider
   */
  registerProvider(provider: ToolProvider): void {
    this.logger.debug(`Registering tool provider: ${provider.name}`);

    this.providers.set(provider.name, provider);

    // Register all tools from the provider
    const tools = provider.getTools();
    for (const tool of tools) {
      this.tools.set(tool.name, tool);
      this.logger.debug(`Registered tool: ${tool.name}`);
    }
  }

  /**
   * Unregisters a tool provider
   */
  unregisterProvider(providerName: string): void {
    this.logger.debug(`Unregistering tool provider: ${providerName}`);

    const provider = this.providers.get(providerName);
    if (!provider) {
      this.logger.warn(`Tool provider not found: ${providerName}`);
      return;
    }

    // Unregister all tools from the provider
    const tools = provider.getTools();
    for (const tool of tools) {
      this.tools.delete(tool.name);
      this.logger.debug(`Unregistered tool: ${tool.name}`);
    }

    this.providers.delete(providerName);
  }

  /**
   * Gets all available tools
   */
  getTools(): Tool[] {
    return Array.from(this.tools.values());
  }

  /**
   * Gets a specific tool by name
   */
  getTool(name: string): Tool | undefined {
    return this.tools.get(name);
  }

  /**
   * Gets all registered providers
   */
  getProviders(): ToolProvider[] {
    return Array.from(this.providers.values());
  }

  /**
   * Checks if a tool is available
   */
  hasTool(name: string): boolean {
    return this.tools.has(name);
  }

  /**
   * Executes a tool call (legacy method without sessionId)
   */
  async executeTool(toolCall: ToolCall): Promise<ToolResult> {
    // Call the new method with undefined sessionId (no tool call reporting)
    return this.executeToolWithSession(toolCall, undefined);
  }

  /**
   * Executes a tool call with session context and tool call reporting
   * Per ACP spec: Reports tool execution via session/update notifications
   */
  async executeToolWithSession(
    toolCall: ToolCall,
    sessionId?: string
  ): Promise<ToolResult> {
    this.logger.debug(`Executing tool: ${toolCall.name}`, {
      id: toolCall.id,
      parameters: toolCall.parameters,
      sessionId,
    });

    const startTime = Date.now();

    // Report tool call if we have a session and tool call manager
    let toolCallId: string | undefined;
    const shouldReportToolCalls = sessionId && this.toolCallManager;

    try {
      const tool = this.tools.get(toolCall.name);
      if (!tool) {
        const duration = Date.now() - startTime;
        return {
          success: false,
          error: `Tool not found: ${toolCall.name}`,
          metadata: {
            toolName: toolCall.name,
            duration,
            executedAt: new Date(),
          },
        };
      }

      // Validate parameters
      const validationError = this.validateToolParameters(
        tool,
        toolCall.parameters
      );
      if (validationError) {
        const duration = Date.now() - startTime;
        return {
          success: false,
          error: `Invalid parameters for ${toolCall.name}: ${validationError}`,
          metadata: {
            toolName: toolCall.name,
            duration,
            executedAt: new Date(),
          },
        };
      }

      // Report tool call start
      if (shouldReportToolCalls) {
        const toolKind = this.getToolKind(toolCall.name);
        const locations = this.extractLocations(toolCall.parameters);
        const reportOptions: {
          title: string;
          kind: ToolKind;
          status: 'in_progress';
          rawInput: Record<string, any>;
          locations?: ToolCallLocation[];
        } = {
          title: this.getToolTitle(toolCall.name, toolCall.parameters),
          kind: toolKind,
          status: 'in_progress',
          rawInput: toolCall.parameters,
        };
        if (locations.length > 0) {
          reportOptions.locations = locations;
        }
        toolCallId = await this.toolCallManager!.reportToolCall(
          sessionId!,
          toolCall.name,
          reportOptions
        );
      }

      // Execute the tool
      const result = await tool.handler(toolCall.parameters);

      const duration = Date.now() - startTime;
      this.logger.debug(`Tool executed in ${duration}ms: ${toolCall.name}`);

      // Report tool call completion
      if (shouldReportToolCalls && toolCallId) {
        if (result.success) {
          // Check if tool result includes diffs (from cursor-tools)
          let content;
          if (
            result.metadata?.['diffs'] &&
            Array.isArray(result.metadata['diffs'])
          ) {
            content = this.toolCallManager!.convertDiffContent(
              result.metadata['diffs']
            );
          }

          await this.toolCallManager!.completeToolCall(sessionId!, toolCallId, {
            rawOutput: result.result,
            ...(content && { content }),
          });
        } else {
          await this.toolCallManager!.failToolCall(sessionId!, toolCallId, {
            error: result.error || 'Unknown error',
            rawOutput: result.result,
          });
        }
      }

      return {
        ...result,
        metadata: {
          ...result.metadata,
          toolName: toolCall.name,
          duration,
          executedAt: new Date(),
          ...(toolCallId && { toolCallId }),
        },
      };
    } catch (error) {
      const duration = Date.now() - startTime;
      this.logger.error(
        `Tool execution failed after ${duration}ms: ${toolCall.name}`,
        error
      );

      const errorMessage =
        error instanceof ToolError
          ? error.message
          : error instanceof Error
            ? error.message
            : String(error);

      // Report tool call failure
      if (shouldReportToolCalls && toolCallId) {
        await this.toolCallManager!.failToolCall(sessionId!, toolCallId, {
          error: errorMessage,
        });
      } else if (shouldReportToolCalls && !toolCallId) {
        // If we didn't get a toolCallId yet, report the failure
        const toolKind = this.getToolKind(toolCall.name);
        const locations = this.extractLocations(toolCall.parameters);
        const reportOptions: {
          title: string;
          kind: ToolKind;
          status: 'failed';
          rawInput: Record<string, any>;
          locations?: ToolCallLocation[];
        } = {
          title: this.getToolTitle(toolCall.name, toolCall.parameters),
          kind: toolKind,
          status: 'failed',
          rawInput: toolCall.parameters,
        };
        if (locations.length > 0) {
          reportOptions.locations = locations;
        }
        toolCallId = await this.toolCallManager!.reportToolCall(
          sessionId!,
          toolCall.name,
          reportOptions
        );
        await this.toolCallManager!.failToolCall(sessionId!, toolCallId, {
          error: errorMessage,
        });
      }

      return {
        success: false,
        error: errorMessage,
        metadata: {
          toolName: toolCall.name,
          duration,
          executedAt: new Date(),
          ...(toolCallId && { toolCallId }),
        },
      };
    }
  }

  /**
   * Extract file locations from tool parameters
   * Per ACP spec: Enable clients to track which files the agent is accessing
   */
  private extractLocations(
    parameters: Record<string, any>
  ): ToolCallLocation[] {
    const locations: ToolCallLocation[] = [];

    // Extract file path from different parameter names
    if (parameters['path']) {
      locations.push({ path: parameters['path'] });
    } else if (parameters['sourcePath']) {
      locations.push({ path: parameters['sourcePath'] });
    }

    // For move/copy operations, also include destination
    if (parameters['destinationPath'] || parameters['destination']) {
      locations.push({
        path: parameters['destinationPath'] || parameters['destination'],
      });
    }

    // For tools that work with multiple files
    if (parameters['files'] && Array.isArray(parameters['files'])) {
      parameters['files'].forEach((file: string) => {
        locations.push({ path: file });
      });
    }

    return locations;
  }

  /**
   * Get the ACP tool kind for a tool name
   */
  private getToolKind(toolName: string): ToolKind {
    // Map tool names to ACP tool kinds
    const kindMap: Record<string, ToolKind> = {
      // Filesystem tools
      read_file: 'read',
      write_file: 'edit',
      list_directory: 'read',
      delete_file: 'delete',
      move_file: 'move',
      copy_file: 'read',
      create_directory: 'edit',

      // Terminal tools
      execute_command: 'execute',
      start_shell_session: 'execute',
      send_to_shell: 'execute',
      kill_shell_session: 'execute',

      // Cursor tools
      search_codebase: 'search',
      analyze_code: 'read',
      apply_code_changes: 'edit',
      run_tests: 'execute',
      get_project_info: 'read',
      explain_code: 'read',
    };

    return kindMap[toolName] || 'other';
  }

  /**
   * Generate a human-readable title for a tool call
   */
  private getToolTitle(
    toolName: string,
    parameters: Record<string, any>
  ): string {
    // Create descriptive titles based on tool and parameters
    switch (toolName) {
      case 'read_file':
        return `Reading file: ${parameters['path'] || 'unknown'}`;
      case 'write_file':
        return `Writing file: ${parameters['path'] || 'unknown'}`;
      case 'list_directory':
        return `Listing directory: ${parameters['path'] || 'unknown'}`;
      case 'delete_file':
        return `Deleting file: ${parameters['path'] || 'unknown'}`;
      case 'move_file':
        return `Moving file: ${parameters['source'] || 'unknown'} → ${parameters['destination'] || 'unknown'}`;
      case 'execute_command':
        return `Executing: ${parameters['command'] || 'unknown command'}`;
      case 'search_codebase':
        return `Searching codebase: ${parameters['query'] || 'unknown'}`;
      case 'analyze_code':
        return `Analyzing: ${parameters['file_path'] || 'unknown'}`;
      case 'apply_code_changes':
        return `Applying ${Array.isArray(parameters['changes']) ? parameters['changes'].length : 0} code changes`;
      case 'run_tests':
        return `Running tests: ${parameters['test_pattern'] || 'all'}`;
      case 'get_project_info':
        return 'Getting project information';
      case 'explain_code':
        return `Explaining code: ${parameters['file_path'] || 'unknown'}`;
      default:
        return `Executing tool: ${toolName}`;
    }
  }

  /**
   * Gets tool capabilities for ACP initialization
   */
  getCapabilities(): Record<string, any> {
    const capabilities: Record<string, any> = {
      tools: Array.from(this.tools.keys()),
      providers: Array.from(this.providers.keys()),
    };

    // Add capability flags based on available tools
    capabilities['filesystem'] =
      this.hasTool('read_file') || this.hasTool('write_file');
    capabilities['terminal'] =
      this.hasTool('execute_command') || this.hasTool('start_shell_session');
    capabilities['cursor'] =
      this.hasTool('search_codebase') || this.hasTool('analyze_code');

    return capabilities;
  }

  /**
   * Gets metrics about tool usage
   */
  getMetrics(): Record<string, any> {
    return {
      totalTools: this.tools.size,
      totalProviders: this.providers.size,
      enabledProviders: Array.from(this.providers.keys()),
      // TODO: Add execution metrics, usage stats, etc.
    };
  }

  /**
   * Validates tool configuration
   */
  validateConfiguration(): string[] {
    const errors: string[] = [];

    // Check if filesystem tools are configured correctly
    if (this.config.tools.filesystem.enabled) {
      if (!this.hasTool('read_file') || !this.hasTool('write_file')) {
        errors.push('Filesystem tools enabled but not properly registered');
      }

      if (this.config.tools.filesystem.allowedPaths.length === 0) {
        errors.push('Filesystem tools enabled but no allowed paths configured');
      }
    }

    // Check if terminal tools are configured correctly
    if (this.config.tools.terminal.enabled) {
      if (!this.hasTool('execute_command')) {
        errors.push('Terminal tools enabled but not properly registered');
      }

      if (this.config.tools.terminal.maxProcesses <= 0) {
        errors.push('Terminal tools enabled but maxProcesses is invalid');
      }
    }

    // Check if cursor tools are configured correctly
    if (this.config.tools.cursor?.enabled !== false) {
      if (!this.hasTool('search_codebase') && !this.hasTool('analyze_code')) {
        errors.push('Cursor tools enabled but not properly registered');
      }
    }

    return errors;
  }

  /**
   * Reloads tool providers based on configuration
   */
  async reload(): Promise<void> {
    this.logger.info('Reloading tool registry');

    try {
      // Clear existing providers and tools
      this.providers.clear();
      this.tools.clear();

      // Re-initialize providers
      this.initializeProviders();

      this.logger.info('Tool registry reloaded successfully');
    } catch (error) {
      this.logger.error('Failed to reload tool registry', error);
      throw new ToolError(
        `Failed to reload tools: ${error instanceof Error ? error.message : String(error)}`,
        undefined,
        error instanceof Error ? error : undefined
      );
    }
  }

  // Private helper methods

  private initializeProviders(): void {
    this.logger.debug('Initializing built-in tool providers');

    // Initialize filesystem tools
    if (this.config.tools.filesystem.enabled) {
      const filesystemProvider = new FilesystemToolProvider(
        this.config,
        this.logger
      );
      this.registerProvider(filesystemProvider);
    }

    // Initialize terminal tools
    if (this.config.tools.terminal.enabled) {
      const terminalProvider = new TerminalToolProvider(
        this.config,
        this.logger
      );
      this.registerProvider(terminalProvider);
    }

    // Initialize cursor-specific tools
    if (this.config.tools.cursor?.enabled !== false) {
      const cursorProvider = new CursorToolsProvider(this.config, this.logger);
      this.registerProvider(cursorProvider);
    }
  }

  private validateToolParameters(
    tool: Tool,
    parameters: Record<string, any>
  ): string | null {
    // Handle null/undefined parameters
    if (parameters === null || parameters === undefined) {
      return 'Parameters are required and must be an object';
    }

    if (typeof parameters !== 'object' || Array.isArray(parameters)) {
      return 'Parameters must be an object';
    }

    // Basic parameter validation
    const required = tool.parameters.required || [];

    for (const param of required) {
      if (!(param in parameters)) {
        return `Missing required parameter: ${param}`;
      }
      // Check if the parameter value is null or undefined
      if (parameters[param] === null || parameters[param] === undefined) {
        return `Parameter ${param} cannot be null or undefined`;
      }
    }

    // TODO: Implement more sophisticated parameter validation
    // - Type checking
    // - Format validation
    // - Range checking
    // - Custom validators

    return null;
  }

  /**
   * Cleanup all tool providers
   */
  async cleanup(): Promise<void> {
    this.logger.debug('Cleaning up tool registry');

    // Call cleanup on each provider that has a cleanup method
    for (const [name, provider] of this.providers) {
      if (typeof (provider as any).cleanup === 'function') {
        try {
          this.logger.debug(`Cleaning up provider: ${name}`);
          await (provider as any).cleanup();
        } catch (error) {
          this.logger.warn(`Failed to cleanup provider ${name}`, error);
        }
      }
    }

    this.logger.debug('Tool registry cleanup completed');
  }
}

File: internal/tools/registry.go
package tools

import (
	"fmt"
	"sort"
	"strings"
	"time"

	"github.com/spjoes/cursor-agent-acp/internal/acp"
	"github.com/spjoes/cursor-agent-acp/internal/client"
	"github.com/spjoes/cursor-agent-acp/internal/config"
	"github.com/spjoes/cursor-agent-acp/internal/cursor"
	"github.com/spjoes/cursor-agent-acp/internal/logging"
	"github.com/spjoes/cursor-agent-acp/internal/toolcall"
)

type Tool struct {
	Name        string
	Description string
	Parameters  map[string]any
	Handler     func(params map[string]any) (acp.ToolResult, error)
}

type ToolProvider interface {
	Name() string
	Description() string
	GetTools() []Tool
	Cleanup() error
}

type ToolCall struct {
	ID         string
	Name       string
	Parameters map[string]any
}

type Registry struct {
	cfg    config.Config
	logger *logging.Logger

	providers map[string]ToolProvider
	tools     map[string]Tool

	cursorBridge *cursor.Bridge
	toolCalls    *toolcall.Manager
}

func NewRegistry(cfg config.Config, logger *logging.Logger, cursorBridge *cursor.Bridge) *Registry {
	r := &Registry{
		cfg:          cfg,
		logger:       logger,
		providers:    map[string]ToolProvider{},
		tools:        map[string]Tool{},
		cursorBridge: cursorBridge,
	}
	r.initializeProviders()
	return r
}

func (r *Registry) SetToolCallManager(manager *toolcall.Manager) {
	r.toolCalls = manager
	r.logger.Debug("ToolCallManager registered with ToolRegistry", nil)
}

func (r *Registry) RegisterProvider(provider ToolProvider) {
	r.logger.Debug("Registering tool provider", map[string]any{"provider": provider.Name()})
	r.providers[provider.Name()] = provider
	for _, t := range provider.GetTools() {
		r.tools[t.Name] = t
		r.logger.Debug("Registered tool", map[string]any{"tool": t.Name})
	}
}

func (r *Registry) UnregisterProvider(providerName string) {
	provider, ok := r.providers[providerName]
	if !ok {
		r.logger.Warn("Tool provider not found", map[string]any{"provider": providerName})
		return
	}
	for _, t := range provider.GetTools() {
		delete(r.tools, t.Name)
	}
	delete(r.providers, providerName)
}

func (r *Registry) ConfigureFilesystemProvider(clientCapabilities map[string]any, fsClient client.FileSystemClient) {
	r.UnregisterProvider("filesystem")
	if !r.cfg.Tools.Filesystem.Enabled {
		return
	}
	provider := NewFilesystemProvider(r.cfg, r.logger, clientCapabilities, fsClient)
	r.RegisterProvider(provider)
}

func (r *Registry) GetTools() []Tool {
	tools := make([]Tool, 0, len(r.tools))
	for _, t := range r.tools {
		tools = append(tools, t)
	}
	return tools
}

func (r *Registry) ToolDescriptors() []acp.ToolDescriptor {
	descriptors := make([]acp.ToolDescriptor, 0, len(r.tools))
	for _, t := range r.tools {
		descriptors = append(descriptors, acp.ToolDescriptor{Name: t.Name, Description: t.Description, Parameters: t.Parameters})
	}
	sort.Slice(descriptors, func(i, j int) bool { return descriptors[i].Name < descriptors[j].Name })
	return descriptors
}

func (r *Registry) GetTool(name string) *Tool {
	tool, ok := r.tools[name]
	if !ok {
		return nil
	}
	copy := tool
	return &copy
}

func (r *Registry) GetProviders() []ToolProvider {
	providers := make([]ToolProvider, 0, len(r.providers))
	for _, p := range r.providers {
		providers = append(providers, p)
	}
	return providers
}

func (r *Registry) HasTool(name string) bool {
	_, ok := r.tools[name]
	return ok
}

func (r *Registry) ExecuteTool(toolCall ToolCall) (acp.ToolResult, error) {
	return r.ExecuteToolWithSession(toolCall, "")
}

func (r *Registry) ExecuteToolWithSession(toolCall ToolCall, sessionID string) (acp.ToolResult, error) {
	start := time.Now()
	tool, ok := r.tools[toolCall.Name]
	if !ok {
		return acp.ToolResult{Success: false, Error: "Tool not found: " + toolCall.Name, Metadata: map[string]any{"toolName": toolCall.Name, "duration": 0, "executedAt": time.Now().UTC()}}, nil
	}

	if toolCall.Parameters == nil {
		toolCall.Parameters = map[string]any{}
	}

	if err := validateToolParameters(tool, toolCall.Parameters); err != nil {
		return acp.ToolResult{Success: false, Error: fmt.Sprintf("Invalid parameters for %s: %s", toolCall.Name, err.Error()), Metadata: map[string]any{"toolName": toolCall.Name, "duration": 0, "executedAt": time.Now().UTC()}}, nil
	}

	var toolCallID string
	if sessionID != "" && r.toolCalls != nil {
		locations := extractLocations(toolCall.Parameters)
		report := map[string]any{
			"title":    toolTitle(toolCall.Name, toolCall.Parameters),
			"kind":     toolKind(toolCall.Name),
			"status":   "pending",
			"rawInput": toolCall.Parameters,
		}
		if len(locations) > 0 {
			report["locations"] = locations
		}
		toolCallID = r.toolCalls.ReportToolCall(sessionID, toolCall.Name, report)
		r.toolCalls.UpdateToolCall(sessionID, toolCallID, map[string]any{"status": "in_progress"})
	}

	params := cloneMap(toolCall.Parameters)
	if sessionID != "" {
		params["_sessionId"] = sessionID
	}

	result, err := tool.Handler(params)
	duration := time.Since(start).Milliseconds()
	if err != nil {
		if sessionID != "" && r.toolCalls != nil && toolCallID != "" {
			r.toolCalls.FailToolCall(sessionID, toolCallID, map[string]any{"error": err.Error()})
		}
		return acp.ToolResult{Success: false, Error: err.Error(), Metadata: map[string]any{"toolName": toolCall.Name, "duration": duration, "executedAt": time.Now().UTC(), "toolCallId": toolCallID}}, nil
	}

	if result.Metadata == nil {
		result.Metadata = map[string]any{}
	}
	result.Metadata["toolName"] = toolCall.Name
	result.Metadata["duration"] = duration
	result.Metadata["executedAt"] = time.Now().UTC()
	if toolCallID != "" {
		result.Metadata["toolCallId"] = toolCallID
	}

	if sessionID != "" && r.toolCalls != nil && toolCallID != "" {
		if result.Success {
			complete := map[string]any{"rawOutput": result.Result}
			if diffs, ok := result.Metadata["diffs"].([]any); ok {
				complete["content"] = r.toolCalls.ConvertDiffContent(diffs)
			}
			r.toolCalls.CompleteToolCall(sessionID, toolCallID, complete)
		} else {
			r.toolCalls.FailToolCall(sessionID, toolCallID, map[string]any{"error": result.Error, "rawOutput": result.Result})
		}
	}

	return result, nil
}

func (r *Registry) GetCapabilities() map[string]any {
	toolNames := make([]string, 0, len(r.tools))
	for name := range r.tools {
		toolNames = append(toolNames, name)
	}
	providerNames := make([]string, 0, len(r.providers))
	for name := range r.providers {
		providerNames = append(providerNames, name)
	}
	cap := map[string]any{
		"tools":     toolNames,
		"providers": providerNames,
	}
	cap["filesystem"] = r.HasTool("read_file") || r.HasTool("write_file")
	cap["cursor"] = r.HasTool("search_codebase") || r.HasTool("analyze_code")
	return cap
}

func (r *Registry) Metrics() map[string]any {
	providers := make([]string, 0, len(r.providers))
	for name := range r.providers {
		providers = append(providers, name)
	}
	return map[string]any{"totalTools": len(r.tools), "totalProviders": len(r.providers), "enabledProviders": providers}
}

func (r *Registry) ValidateConfiguration() []string {
	errs := make([]string, 0)
	if r.cfg.Tools.Filesystem.Enabled {
		if _, ok := r.providers["filesystem"]; !ok {
			errs = append(errs, "Filesystem tools enabled but filesystem provider not registered")
		}
	}
	if r.cfg.Tools.Terminal.Enabled && r.cfg.Tools.Terminal.MaxProcesses <= 0 {
		errs = append(errs, "Terminal enabled but maxProcesses is invalid")
	}
	if r.cfg.Tools.Cursor.Enabled && !r.HasTool("search_codebase") && !r.HasTool("analyze_code") {
		errs = append(errs, "Cursor tools enabled but not properly registered")
	}
	return errs
}

func (r *Registry) Reload() error {
	for _, provider := range r.providers {
		_ = provider.Cleanup()
	}
	r.providers = map[string]ToolProvider{}
	r.tools = map[string]Tool{}
	r.initializeProviders()
	return nil
}

func (r *Registry) Cleanup() error {
	for _, provider := range r.providers {
		if err := provider.Cleanup(); err != nil {
			r.logger.Warn("Failed to cleanup provider", map[string]any{"provider": provider.Name(), "error": err.Error()})
		}
	}
	return nil
}

func (r *Registry) initializeProviders() {
	if r.cfg.Tools.Cursor.Enabled {
		r.RegisterProvider(NewCursorProvider(r.cfg, r.logger, r.cursorBridge))
	}
}

func validateToolParameters(tool Tool, params map[string]any) error {
	if params == nil {
		return fmt.Errorf("parameters are required and must be an object")
	}
	required, _ := tool.Parameters["required"].([]string)
	if len(required) == 0 {
		if raw, ok := tool.Parameters["required"].([]any); ok {
			required = make([]string, 0, len(raw))
			for _, v := range raw {
				required = append(required, fmt.Sprint(v))
			}
		}
	}
	for _, key := range required {
		value, ok := params[key]
		if !ok || value == nil {
			return fmt.Errorf("Missing required parameter: %s", key)
		}
	}
	return nil
}

func extractLocations(parameters map[string]any) []map[string]any {
	locations := make([]map[string]any, 0)
	if path, ok := parameters["path"]; ok {
		locations = append(locations, map[string]any{"path": path})
	} else if source, ok := parameters["sourcePath"]; ok {
		locations = append(locations, map[string]any{"path": source})
	}
	if dest, ok := parameters["destinationPath"]; ok {
		locations = append(locations, map[string]any{"path": dest})
	} else if dest, ok := parameters["destination"]; ok {
		locations = append(locations, map[string]any{"path": dest})
	}
	if files, ok := parameters["files"].([]any); ok {
		for _, file := range files {
			locations = append(locations, map[string]any{"path": file})
		}
	}
	return locations
}

func toolKind(name string) string {
	kindMap := map[string]string{
		"read_file": "read", "copy_file": "read", "list_directory": "read", "get_file_info": "read",
		"write_file": "edit", "append_file": "edit", "create_file": "edit", "patch_file": "edit", "apply_code_changes": "edit",
		"delete_file": "delete", "remove_file": "delete", "remove_directory": "delete",
		"move_file": "move", "rename_file": "move",
		"search_codebase": "search", "search_files": "search", "grep": "search", "find_files": "search", "find_references": "search", "find_definitions": "search",
		"run_tests": "execute", "run_command": "execute", "execute_command": "execute", "run_script": "execute", "shell": "execute",
		"fetch_url": "fetch", "http_request": "fetch", "download_file": "fetch", "api_request": "fetch", "web_search": "fetch",
		"think": "think", "reason": "think", "plan": "think", "analyze": "think", "explain_code": "think",
		"switch_mode": "switch_mode", "set_mode": "switch_mode", "change_mode": "switch_mode",
		"analyze_code": "read", "get_project_info": "read",
	}
	if kind, ok := kindMap[name]; ok {
		return kind
	}
	return "other"
}

func toolTitle(toolName string, parameters map[string]any) string {
	switch toolName {
	case "read_file":
		return "Reading file: " + str(parameters["path"], "unknown")
	case "write_file":
		return "Writing file: " + str(parameters["path"], "unknown")
	case "list_directory":
		return "Listing directory: " + str(parameters["path"], "unknown")
	case "delete_file", "remove_file":
		return "Deleting file: " + str(parameters["path"], "unknown")
	case "remove_directory":
		return "Removing directory: " + str(parameters["path"], "unknown")
	case "move_file", "rename_file":
		return "Moving file: " + str(parameters["source"], str(parameters["from"], "unknown")) + " -> " + str(parameters["destination"], str(parameters["to"], "unknown"))
	case "search_codebase":
		return "Searching codebase: " + str(parameters["query"], "unknown")
	case "run_tests":
		return "Running tests: " + str(parameters["test_pattern"], "all")
	case "run_command", "execute_command", "shell":
		return "Running: " + str(parameters["command"], "unknown")
	case "analyze", "analyze_code":
		return "Analyzing: " + str(parameters["file_path"], str(parameters["target"], "unknown"))
	case "get_project_info":
		return "Getting project information"
	case "explain_code":
		return "Explaining code: " + str(parameters["file_path"], "unknown")
	default:
		return "Executing tool: " + toolName
	}
}

func str(v any, def string) string {
	if v == nil {
		return def
	}
	s := strings.TrimSpace(fmt.Sprint(v))
	if s == "" {
		return def
	}
	return s
}

func cloneMap(in map[string]any) map[string]any {
	out := make(map[string]any, len(in))
	for k, v := range in {
		out[k] = v
	}
	return out
}
